File: FirstAkkaApplication/src/main/java/akka/first/app/java/actors/ReduceActor.java
package akka.first.app.java.actors;

import java.util.HashMap;
import java.util.List;

import akka.actor.ActorRef;
import akka.actor.UntypedActor;
import akka.first.app.java.messages.MapData;
import akka.first.app.java.messages.ReduceData;
import akka.first.app.java.messages.WordCount;

public class ReduceActor extends UntypedActor {

	private ActorRef aggregateActor = null;

	public ReduceActor(ActorRef inAggregateActor) {
		aggregateActor = inAggregateActor;
	}

	@Override
	public void onReceive(Object message) throws Exception {
		if (message instanceof MapData) {
			MapData mapData = (MapData) message;
			// reduce the incoming data
			ReduceData reduceData = reduce(mapData.getDataList());
			// forward the result to aggregate actor
			aggregateActor.tell(reduceData);
		} else
			unhandled(message);
	}

	private ReduceData reduce(List<WordCount> dataList) {
		HashMap<String, Integer> reducedMap = new HashMap<String, Integer>();
		for (WordCount wordCount : dataList) {
			if (reducedMap.containsKey(wordCount.getWord())) {
				Integer value = (Integer) reducedMap.get(wordCount.getWord());
				value++;
				reducedMap.put(wordCount.getWord(), value);
			} else {
				reducedMap.put(wordCount.getWord(), Integer.valueOf(1));
			}
		}
		return new ReduceData(reducedMap);
	}
}

File: FirstAkkaApplication/src/main/scala/akka/first/app/scala/actors/ReduceActor.scala
package akka.first.app.scala.actors

import scala.collection.immutable._

import akka.actor.actorRef2Scala
import akka.actor.Actor
import akka.actor.ActorRef
import akka.first.app.scala.MapData
import akka.first.app.scala.ReduceData
import akka.first.app.scala.Word

class ReduceActor(aggregateActor: ActorRef) extends Actor {

  val defaultCount: Int = 1
  def receive: Receive = {
    case message: MapData =>
      aggregateActor ! reduce(message.dataList)
  }

  def reduce(dataList: List[Word]): ReduceData = {
    var reducedMap = new HashMap[String, Int]
    for (wc: Word <- dataList) {
      var word: String = wc.word
      if (reducedMap.contains(word)) {  
    	var count:Int = reducedMap.get(word).get + defaultCount  
        reducedMap += word -> count
      } else {
        reducedMap += word -> defaultCount
      }
    }
    return new ReduceData(reducedMap)
  }
}